collectors/collector.go

```diff
@@ -0,0 +1,165 @@
+package collectors
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"time"
+)
+
+// Collector is the interface to abstract data collection from different sources
+type Collector interface {
+	// Name returns the Name of this collector
+	Name() string
+	// Optional returns true if Collector is not mandatory and fails should be ignored
+	Optional() bool
+	// Collect returns collected data
+	Collect(ctx context.Context) (io.ReadCloser, error)
+}
+
+type NameOptional struct {
```

Comment:
Let's not do this, please. It saves us some lines of code but comes with the cost of decreased readibility. Let's just add `Name` and `Optional` explicitly to the structs used as `Collectors`.

Reply:
I added it to prevent name clash on `Name()` function and `Name` field. Can you propose new "names" for that `Names`?

Reply:
Can't you just use lowercase names for the fields?

Reply:
How we will initialize them? Should we have a factory method for all collectors?

Reply:
I actually like the Go pattern of having `NewXYZ` functions. So yes, we could do it.

Reply:
FYI - Here's the recommendation in [Effective Go](https://golang.org/doc/effective_go.html#composite_literals)